src/sign/mod.rs

```diff
@@ -1,10 +1,387 @@
 //!
 //! **This module is experimental and likely to change significantly.**
-#![cfg(feature = "sign")]
-#![cfg_attr(docsrs, doc(cfg(feature = "sign")))]
+//!
+//! Signatures are at the heart of DNSSEC -- they confirm the authenticity of
+//! a DNS record served by a security-aware name server.  Signatures can be
+//! made "online" (in an authoritative name server while it is running) or
+//! "offline" (outside of a name server).  Once generated, signatures can be
+//! serialized as DNS records and stored alongside the authenticated records.
+//!
+//! A DNSSEC key actually has two components: a cryptographic key, which can
+//! be used to make and verify signatures, and key metadata, which defines how
+//! the key should be used.  These components are brought together by the
+//! [`SigningKey`] type.  It must be instantiated with a cryptographic key
+//! type, such as [`common::KeyPair`], in order to be used.
+//!
+//! # Example Usage
+//!
+//! At the moment, only "low-level" signing is supported.
+//!
+//! ```
+//! # use domain::sign::*;
+//! # use domain::base::Name;
+//! // Generate a new ED25519 key.
+//! let params = GenerateParams::Ed25519;
+//! let (sec_bytes, pub_bytes) = common::generate(params).unwrap();
+//!
+//! // Parse the key into Ring or OpenSSL.
+//! let key_pair = common::KeyPair::from_bytes(&sec_bytes, &pub_bytes).unwrap();
+//!
+//! // Associate the key with important metadata.
+//! let owner: Name<Vec<u8>> = "www.example.org.".parse().unwrap();
+//! let flags = 257; // key signing key
+//! let key = SigningKey::new(owner, flags, key_pair);
+//!
+//! // Access the public key (with metadata).
+//! let pub_key = key.public_key();
+//! println!("{:?}", pub_key);
+//!
+//! // Sign arbitrary byte sequences with the key.
+//! let sig = key.raw_secret_key().sign_raw(b"Hello, World!").unwrap();
+//! println!("{:?}", sig);
+//! ```
+//!
+//! # Cryptography
+//!
+//! This crate supports OpenSSL and Ring for performing cryptography.  These
+//! cryptographic backends are gated on the `openssl` and `ring` features,
+//! respectively.  They offer mostly equivalent functionality, but OpenSSL
+//! supports a larger set of signing algorithms.  A [`common`] backend is
+//! provided for users that wish to use either or both backends at runtime.
+//!
+//! Each backend module exposes a `KeyPair` type, representing a cryptographic
+//! key that can be used for signing, and a `generate()` function for creating
+//! new keys.
+//!
+//! Users can choose to bring their own cryptography by providing their own
+//! `KeyPair` type that implements [`SignRaw`].  Note that `async` signing
+//! (useful for interacting with cryptographic hardware like HSMs) is not
+//! currently supported.
+//!
+//! While each cryptographic backend can support a limited number of signature
+//! algorithms, even the types independent of a cryptographic backend (e.g.
+//! [`SecretKeyBytes`] and [`GenerateParams`]) support a limited number of
+//! algorithms.  They are:
```

Comment:
I wanted some module-level documentation so that readers don't need to jump between declarations to get a complete high-level understanding of the module.  I guess it's a balance between concision and readability.

Reply:
I'd be worried about the potential for it becoming out of sync with the underlying implementation.

Reply:
That's understandable.  I'll fix this.